core/server/master/src/main/java/alluxio/master/file/mdsync/DefaultSyncProcess.java

```diff
@@ -0,0 +1,997 @@
+        handleConcurrentModification(
+            syncState.mContext, currentInode.getLockedPath().getUri().getPath(), false, e);
+      }
+      return new SingleInodeSyncResult(false, true, true);
+    }
+    // (Case 4) - in this case both the inode, and the UFS item exist, so we check if we need
+    // to update the metadata
+    LockedInodePath lockedInodePath = currentInode.getLockedPath();
+    lockedInodePath.traverse();
+    // skip if this is a mount point
+    if (mMountTable.isMountPoint(currentInode.getLockedPath().getUri())) {
+      syncState.mContext.reportSyncOperationSuccess(SyncOperation.SKIPPED_ON_MOUNT_POINT, 1);
+      return new SingleInodeSyncResult(true, true, true);
+    }
+    // skip if the file is not complete or not persisted
+    if (lockedInodePath.getInode().isFile()) {
+      InodeFile inodeFile = lockedInodePath.getInodeFile();
+      if (!inodeFile.isCompleted() || !inodeFile.isPersisted()) {
+        syncState.mContext.reportSyncOperationSuccess(SyncOperation.SKIPPED_NON_PERSISTED, 1);
+        return new SingleInodeSyncResult(true, true, false);
+      }
+    }
+    // HDFS also fetches ACL list, which is ignored for now
+    String ufsType = syncState.mUfs.getUnderFSType();
+    Fingerprint ufsFingerprint = Fingerprint.create(ufsType, currentUfsStatus.mUfsItem);
+    boolean containsMountPoint = mMountTable.containsMountPoint(
+        currentInode.getLockedPath().getUri(), true, false);
+    UfsSyncUtils.SyncPlan syncPlan =
+        UfsSyncUtils.computeSyncPlan(currentInode.getInode(), ufsFingerprint, containsMountPoint);
+    if (syncPlan.toUpdateMetaData() || syncPlan.toDelete() || syncPlan.toLoadMetadata()) {
+      try {
+        if (syncPlan.toUpdateMetaData()) {
+          /*
+            Inode obtained from the iterator might be stale, especially if the inode is persisted
+            in the rocksDB. This in generally should not cause issues. But if the sync plan is to
+            update the inode, the inode reference might not be the inode we want to update, and we
+            might do wrong updates in an incorrect inode.
+            e.g. considering the following steps:
+              1. create a file /file in alluxio, the file is persisted in UFS
+              2. change the MODE of /file, without applying this change in UFS
+              3. do a metadata sync on /file, as the fingerprints are different,
+                 /file is expected to be updated using the metadata from UFS
+              4. during the metadata sync, the file /file is deleted from alluxio, a DIRECTORY
+                 with the same name /file is created
+              5. the metadata of /file in UFS will be used to update the directory /file, which
+                 is incorrect.
+              To prevent this from happening, we re-fetch the inode from the locked inode path
+              and makes sure the reference is the same, before we update the inode.
+           */
```

Comment:
I will remove this check since, we lock before checking the inode now. This should be ok right @elega

Reply:
yeah this should be okay... my concern is that previously we could have avoided unnecessary locking if we DONOT lock before the check and only locks when the actual update is needed and now we won't be able to do that any more.
But the previous optimization is also a bit aggressive to me as it might create some concurrency issues. 

What's your thoughts? Why did you change the locking pattern in the first place? 
Regardless I think this code can be removed now and we don't make to make any big logic change as we are closing to the merge.